test/unit/puppetlabs/services/jruby_pool_manager/jruby_ref_pool_test.clj

```diff
@@ -12,41 +12,48 @@
                                   :multithreaded true
                                   :max-borrows-per-instance max-borrows})))
 
+(deftest borrow-while-no-instances-available-test
+  (testing "when all instances are in use, borrow blocks until an instance becomes available"
+    (let [pool-size 2]
+      (jruby-testutils/with-pool-context
+        pool-context
+        jruby-testutils/default-services
+        (jruby-testutils/jruby-config {:max-active-instances pool-size
+                                       :multithreaded true})
+        ;; borrow both instances from the pool
+        (let [drained-instances (jruby-testutils/drain-pool pool-context pool-size)]
+          (try
+            (is (= 2 (count drained-instances)))
+
+            ;; attempt a borrow, which will block because no instances are free
+            (let [borrow-instance (future (jruby-core/borrow-from-pool-with-timeout
+                                             pool-context
+                                             :borrow-with-no-free-instances-test
+                                             []))]
+              (is (not (realized? borrow-instance)))
+
+              ;; return an instance to the pool
+              (jruby-core/return-to-pool pool-context
+                                          (first drained-instances)
+                                          :borrow-with-no-free-instances-test
+                                          [])
+
+              ;; now the borrow can complete
+              (is (some? @borrow-instance)))
+            (finally
+              (jruby-testutils/fill-drained-pool pool-context drained-instances))))))))
```

Comment:
Why do we need to refill after the test?

Reply:
Because we need to make sure that we can shutdown as the last step of the macro, so that all the JRuby containers get shutdown properly (see https://github.com/puppetlabs/jruby-utils/commit/b08a9b0fb8c6df62a11da98e97dec9b10ef289ec). The shutdown can't proceed if we're holding onto any references.